File: 2020-algorithms-studying/Trees/BinaryTreeRightSideView.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
    /**
    We want to traverse tree and return the right side nodes
    
    **/
    const cache = []
    const traverseRight = ( node , level ) =>{
        //As we traverse each level we want to push each rightmost element into the cache
        if(node === null) return
        if(level == cache.length){
            cache.push(node.val)
        }
        if(node.right !== null){
            traverseRight(node.right, level + 1)
        }
        if(node.left !== null){
            traverseRight(node.left, level + 1)
        }
    }
    traverseRight(root, 0);
    return cache
};

File: binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        
        
        queue = deque([root])
        
        
        while queue:
            level = len(queue)
            for i in range(level):
                curr = queue.popleft()
                if i == level - 1:
                    yield curr.val
                if curr.left: queue.append(curr.left)
                if curr.right: queue.append(curr.right)
        